### PR TITLE
Avoid a few locks

### PR DESCRIPTION
**Pull request by oli-obk**
We can use atomics or datastructures tuned for specific access patterns instead of locks. This may be an improvement for parallel rustc, but it's mostly a cleanup making various datastructures only usable in the way they are used right now (append data, never mutate), instead of having a general purpose lock.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Sounds good! I'll try review within the next couple of days.

File: compiler/rustc_data_structures/src/sync/vec.rs [hunk 1]
```diff
@@ -39,3 +39,35 @@ impl<I: Idx, T: Copy> AppendOnlyVec<I, T> {
...
+        #[cfg(parallel_compiler)]
+        self.vec.push(val)
+    }
+
+    pub fn get(&self, i: usize) -> Option<T> {
+        #[cfg(not(parallel_compiler))]
+        return self.vec.get_copy(i);
+        #[cfg(parallel_compiler)]
+        return self.vec.get(i);
+    }
```
Comment (michaelwoerister) [inline compiler/rustc_data_structures/src/sync/vec.rs:77]: This file would be a good candidate for `cfg_if!` if you want to clean it up.

File: compiler/rustc_session/src/parse.rs [hunk 2]
```diff
@@ -324,13 +324,16 @@ impl ParseSess {
...
 
-    pub fn proc_macro_quoted_spans(&self) -> Vec<Span> {
-        self.proc_macro_quoted_spans.lock().clone()
+    pub fn proc_macro_quoted_spans(&self) -> impl Iterator<Item = (usize, Span)> + '_ {
+        // This is equivalent to `.iter().copied().enumerate()`, but that isn't possible for
+        // AppendOnlyVec, so we resort to this scheme.
+        (0..)
+            .map(|i| (i, self.proc_macro_quoted_spans.get(i)))
+            .take_while(|(_, o)| o.is_some())
+            .filter_map(|(i, o)| Some((i, o?)))
```
Comment (michaelwoerister) [inline compiler/rustc_session/src/parse.rs:336]: Can you replace this with `iter_enumerated` that's been added in a later commit?

**Review by michaelwoerister — COMMENTED** [review]
Looks good to me! I've left a few comments below.

File: compiler/rustc_data_structures/src/sync/vec.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_data_structures/src/sync/vec.rs:77]: It doesn't actually make anything simpler as we don't allow attributes on expressions and we can't use cfg_if on fields or their types.

File: compiler/rustc_data_structures/src/sync.rs [hunk 3]
```diff
@@ -82,6 +82,14 @@ cfg_if! {
             }
         }
 
+        impl Atomic<bool> {
+            pub fn fetch_or(&self, val: bool, _: Ordering) -> bool {
+                let result = self.0.get() | val;
+                self.0.set(val);
+                result
+            }
```
Comment (krdln) [inline compiler/rustc_data_structures/src/sync.rs:90]: I think this is not equivalent of AtomicBools' fetch_or. Perhaps we want the following?

```suggestion
            pub fn fetch_or(&self, val: bool, _: Ordering) -> bool {
                let old = self.0.get();
                self.0.set(old | val);
                old
            }
```

**Comment by oli-obk** [comment]
I can reproduce the segfault locally in parallel-rustc. Investigating...

**Comment by oli-obk** [comment]
I fixed the issue upstream in `elsa` and updated our dependency to pick up the new version.

@bors r=michaelwoerister

**Pull request merged**